// igniteui-angular/esm2022/lib/calendar/common/calendar-view.directive.mjs
import { Output, EventEmitter, Input, HostListener, ViewChildren, booleanAttribute, Directive, HostBinding, InjectionToken, Inject, } from "@angular/core";
import { noop } from "rxjs";
import { NG_VALUE_ACCESSOR } from "@angular/forms";
import { IGX_CALENDAR_VIEW_ITEM, } from "../calendar.directives";
import { CalendarDay } from "../common/model";
import { getNextActiveDate, isDateInRanges } from "./helpers";
import { DateRangeType } from "../../core/dates";
import * as i0 from "@angular/core";
export var Direction;
(function (Direction) {
    Direction[Direction["NEXT"] = 1] = "NEXT";
    Direction[Direction["PREV"] = -1] = "PREV";
})(Direction || (Direction = {}));
export const DAY_INTERVAL_TOKEN = new InjectionToken("DAY_INTERVAL");
export class IgxCalendarViewDirective {
    get activeDescendant() {
        if (this.tabIndex === -1)
            return;
        return this.date.getTime();
    }
    /**
     * Gets/sets the selected date of the view.
     * By default it's the current date.
     * ```typescript
     * let date = this.view.date;
     * ```
     *
     * @memberof IgxYearsViewComponent
     */
    set date(value) {
        if (!(value instanceof Date))
            return;
        this._date = value;
    }
    get date() {
        return this._date;
    }
    /**
     * Gets the `locale` of the view.
     * Default value is `"en"`.
     * ```typescript
     * let locale = this.view.locale;
     * ```
     *
     * @memberof IgxCalendarViewDirective
     */
    get locale() {
        return this._locale;
    }
    /**
     * Sets the `locale` of the view.
     * Expects a valid BCP 47 language tag.
     * Default value is `"en"`.
     *
     * @memberof IgxCalendarViewDirective
     */
    set locale(value) {
        this._locale = value;
        this.initFormatter();
    }
    constructor(dayInterval) {
        this.dayInterval = dayInterval;
        this.role = 'grid';
        this.tabIndex = 0;
        /**
         * Applies styles to the active item on view focus.
         */
        this.showActive = false;
        /**
         * Emits an event when a selection is made in the view.
         * Provides reference the `date` property in the component.
         * @memberof IgxCalendarViewDirective
         */
        this.selected = new EventEmitter();
        /**
         * Emits an event when a page changes in the view.
         * Provides reference the `date` property in the component.
         * @memberof IgxCalendarViewDirective
         * @hidden @internal
         */
        this.pageChanged = new EventEmitter();
        /**
         * Emits an event when the active date has changed.
         * @memberof IgxCalendarViewDirective
         * @hidden @internal
         */
        this.activeDateChanged = new EventEmitter();
        /**
         * @hidden
         */
        this._locale = "en";
        /**
         * @hidden
         * @internal
         */
        this._date = new Date();
        /**
         * @hidden
         */
        this._onTouchedCallback = noop;
        /**
         * @hidden
         */
        this._onChangeCallback = noop;
        this.initFormatter();
    }
    /**
     * @hidden
     */
    onKeydownArrowDown(event) {
        this.navigateTo(event, Direction.NEXT, 3);
    }
    /**
     * @hidden
     */
    onKeydownArrowUp(event) {
        this.navigateTo(event, Direction.PREV, 3);
    }
    /**
     * @hidden
     */
    onKeydownArrowRight(event) {
        this.navigateTo(event, Direction.NEXT, 1);
    }
    /**
     * @hidden
     */
    onKeydownArrowLeft(event) {
        this.navigateTo(event, Direction.PREV, 1);
    }
    /**
     * @hidden
     */
    onKeydownHome(event) {
        event.preventDefault();
        event.stopPropagation();
        this.date = this.range.at(0);
        this.activeDateChanged.emit(this.date);
    }
    /**
     * @hidden
     */
    onKeydownEnd(event) {
        event.preventDefault();
        event.stopPropagation();
        this.date = this.range.at(-1);
        this.activeDateChanged.emit(this.date);
    }
    /**
     * @hidden
     */
    onKeydownEnter(event) {
        event.stopPropagation();
        this.selected.emit(this.date);
        this._onChangeCallback(this.date);
    }
    /**
     * @hidden
     */
    handleFocus() {
        this.showActive = true;
    }
    /**
     * @hidden
     */
    handleBlur() {
        this.showActive = false;
    }
    /**
     * @hidden
     */
    selectDate(value) {
        this.date = value;
        this.selected.emit(this.date);
        this._onChangeCallback(this.date);
    }
    /**
     * @hidden
     */
    registerOnChange(fn) {
        this._onChangeCallback = fn;
    }
    /**
     * @hidden
     */
    registerOnTouched(fn) {
        this._onTouchedCallback = fn;
    }
    /**
     * @hidden
     */
    writeValue(value) {
        if (value) {
            this.date = value;
        }
    }
    /**
     * @hidden
     */
    navigateTo(event, direction, delta) {
        event.preventDefault();
        event.stopPropagation();
        const date = getNextActiveDate(CalendarDay.from(this.date).add(this.dayInterval, direction * delta), []);
        const outOfRange = !isDateInRanges(date, [
            {
                type: DateRangeType.Between,
                dateRange: [this.range.at(0), this.range.at(-1)],
            },
        ]);
        if (outOfRange) {
            this.pageChanged.emit(date.native);
        }
        this.date = date.native;
        this.activeDateChanged.emit(this.date);
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "18.2.4", ngImport: i0, type: IgxCalendarViewDirective, deps: [{ token: DAY_INTERVAL_TOKEN }], target: i0.ɵɵFactoryTarget.Directive }); }
    static { this.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "16.1.0", version: "18.2.4", type: IgxCalendarViewDirective, isStandalone: true, inputs: { role: "role", tabIndex: "tabIndex", formatView: ["formatView", "formatView", booleanAttribute], showActive: ["showActive", "showActive", booleanAttribute], date: "date", locale: "locale" }, outputs: { selected: "selected", pageChanged: "pageChanged", activeDateChanged: "activeDateChanged" }, host: { listeners: { "keydown.arrowdown": "onKeydownArrowDown($event)", "keydown.arrowup": "onKeydownArrowUp($event)", "keydown.arrowright": "onKeydownArrowRight($event)", "keydown.arrowleft": "onKeydownArrowLeft($event)", "keydown.home": "onKeydownHome($event)", "keydown.end": "onKeydownEnd($event)", "keydown.enter": "onKeydownEnter($event)", "focus": "handleFocus()", "blur": "handleBlur()" }, properties: { "attr.role": "this.role", "attr.tabIndex": "this.tabIndex", "attr.aria-activeDescendant": "this.activeDescendant" } }, providers: [
            {
                provide: NG_VALUE_ACCESSOR,
                useExisting: IgxCalendarViewDirective,
                multi: true,
            },
        ], viewQueries: [{ propertyName: "viewItems", predicate: IGX_CALENDAR_VIEW_ITEM, descendants: true, read: IGX_CALENDAR_VIEW_ITEM }], ngImport: i0 }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "18.2.4", ngImport: i0, type: IgxCalendarViewDirective, decorators: [{
            type: Directive,
            args: [{
                    providers: [
                        {
                            provide: NG_VALUE_ACCESSOR,
                            useExisting: IgxCalendarViewDirective,
                            multi: true,
                        },
                    ],
                    standalone: true,
                }]
        }], ctorParameters: () => [{ type: undefined, decorators: [{
                    type: Inject,
                    args: [DAY_INTERVAL_TOKEN]
                }] }], propDecorators: { role: [{
                type: HostBinding,
                args: ["attr.role"]
            }, {
                type: Input
            }], tabIndex: [{
                type: HostBinding,
                args: ["attr.tabIndex"]
            }, {
                type: Input
            }], activeDescendant: [{
                type: HostBinding,
                args: ['attr.aria-activeDescendant']
            }], formatView: [{
                type: Input,
                args: [{ transform: booleanAttribute }]
            }], showActive: [{
                type: Input,
                args: [{ transform: booleanAttribute }]
            }], selected: [{
                type: Output
            }], pageChanged: [{
                type: Output
            }], activeDateChanged: [{
                type: Output
            }], viewItems: [{
                type: ViewChildren,
                args: [IGX_CALENDAR_VIEW_ITEM, { read: IGX_CALENDAR_VIEW_ITEM }]
            }], date: [{
                type: Input
            }], locale: [{
                type: Input
            }], onKeydownArrowDown: [{
                type: HostListener,
                args: ["keydown.arrowdown", ["$event"]]
            }], onKeydownArrowUp: [{
                type: HostListener,
                args: ["keydown.arrowup", ["$event"]]
            }], onKeydownArrowRight: [{
                type: HostListener,
                args: ["keydown.arrowright", ["$event"]]
            }], onKeydownArrowLeft: [{
                type: HostListener,
                args: ["keydown.arrowleft", ["$event"]]
            }], onKeydownHome: [{
                type: HostListener,
                args: ["keydown.home", ["$event"]]
            }], onKeydownEnd: [{
                type: HostListener,
                args: ["keydown.end", ["$event"]]
            }], onKeydownEnter: [{
                type: HostListener,
                args: ["keydown.enter", ["$event"]]
            }], handleFocus: [{
                type: HostListener,
                args: ["focus"]
            }], handleBlur: [{
                type: HostListener,
                args: ["blur"]
            }] } });
