// igniteui-angular/esm2022/lib/services/theme/theme.service.mjs
import { Inject, Injectable } from "@angular/core";
import { mkenum } from "../../core/utils";
import { BehaviorSubject } from "rxjs";
import { DOCUMENT } from "@angular/common";
import * as i0 from "@angular/core";
const Theme = /*@__PURE__*/ mkenum({
    Material: "material",
    Fluent: "fluent",
    Bootstrap: "bootstrap",
    IndigoDesign: "indigo",
});
export class ThemeService {
    constructor(document) {
        this.document = document;
        this.theme$ = new BehaviorSubject("material");
        this.theme$.asObservable().subscribe((value) => {
            this.globalTheme = value;
        });
        this.init();
    }
    init() {
        const theme = globalThis.window
            ?.getComputedStyle(this.document.body)
            .getPropertyValue("--ig-theme")
            .trim();
        if (theme !== "") {
            this.theme$.next(theme);
        }
    }
    getComponentTheme(el) {
        return globalThis.window
            ?.getComputedStyle(el.nativeElement)
            .getPropertyValue('--theme')
            .trim();
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "18.2.4", ngImport: i0, type: ThemeService, deps: [{ token: DOCUMENT }], target: i0.ɵɵFactoryTarget.Injectable }); }
    static { this.ɵprov = i0.ɵɵngDeclareInjectable({ minVersion: "12.0.0", version: "18.2.4", ngImport: i0, type: ThemeService, providedIn: "root" }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "18.2.4", ngImport: i0, type: ThemeService, decorators: [{
            type: Injectable,
            args: [{
                    providedIn: "root",
                }]
        }], ctorParameters: () => [{ type: undefined, decorators: [{
                    type: Inject,
                    args: [DOCUMENT]
                }] }] });
